added cypress tests for shifting request

Ayush1404 opened the pull request:
## Proposed Changes

- Fixes #8914 
- Added ShiftingRequest.cy.ts file and the cypress tests for the shifting request in  it.
- Created some functions that were useful in context of  shifting request creation and update.

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a suite of end-to-end tests for managing shifting requests.
	- Added functionality for shifting patients and managing ambulance-related information.
	- New classes for handling shift details and updates, including methods for interacting with shift forms.

- **Accessibility Improvements**
	- Enhanced accessibility in the ShiftDetails component by adding unique `id` attributes to various elements.

- **Bug Fixes**
	- Improved stability of tests by ensuring consistent local storage state before and after each test.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Ayush1404 committed:
added cypress tests for shifting request

nihal467 commented on cypress/e2e/shifting_spec/ShiftingRequest.cy.ts:
```diff
@@ -0,0 +1,80 @@
+  const patientConsultationPage = new PatientConsultationPage() 
+
+  before(() => { 
+    loginPage.loginAsDisctrictAdmin(); 
```
```suggestion
    loginPage.loginAsDistrictAdmin(); 
```

nihal467 committed:
Update cypress/e2e/shifting_spec/ShiftingRequest.cy.ts

nihal467 commented on cypress/e2e/shifting_spec/ShiftingRequest.cy.ts:
```diff
@@ -0,0 +1,80 @@
+  });
+
+  it("Creating and verifying new shifting request", () => {
+    cy.intercept("POST", "api/v1/shift").as("shiftRequest"); 
```
- this intercept should be a function
- it should be placed before submitting the form

nihal467 commented on cypress/e2e/shifting_spec/ShiftingRequest.cy.ts:
```diff
@@ -0,0 +1,80 @@
+    cy.intercept("POST", "api/v1/shift").as("shiftRequest"); 
+  
+    cy.visit('/patients')
+    patientPage.visitPatient("Dummy Patient 16")
```
- move all user inputs from 31 to 40 lines, to a constant

nihal467 commented on cypress/e2e/shifting_spec/ShiftingRequest.cy.ts:
```diff
@@ -0,0 +1,80 @@
+    cy.contains("Dummy Patient 16").should("exist");
+  });
+
+  it("Editing and verifying refelctions in existing request", () => {
```
- combine this with the above test itself, because they are dependent test, and test files are executed parallel randomly

nihal467 commented on cypress/e2e/shifting_spec/ShiftingRequest.cy.ts:
```diff
@@ -0,0 +1,80 @@
+    shiftingBoard.openDetails("Dummy Patient 16")
+    shiftDetails.clickUpdateStatusButton()
+
+    shiftUpdate.typeShiftReason("new reason");
```
- declare all input as const and use to across the test

nihal467 commented on cypress/e2e/shifting_spec/ShiftingRequest.cy.ts:
```diff
@@ -0,0 +1,80 @@
+    cy.wait('@shiftUpdateRequest').then((interception) => {
+      const responseData = interception.response.body;
+      
+      expect(responseData.patient_object.name).to.eq("Dummy Patient 16");
```
- verify the modification is reflected in the shifting details page, not just the body of the request

nihal467 commented on cypress/e2e/shifting_spec/ShiftingRequest.cy.ts:
```diff
@@ -0,0 +1,80 @@
+    cy.contains("Dummy Patient 16").should("exist");
+
+    shiftingBoard.openDetails("Dummy Patient 16")
+    shiftDetails.clickUpdateStatusButton()
```
- once you open the view details page, verify the initial data you inputted while creating the form are reflected properly

nihal467 commented on cypress/e2e/shifting_spec/ShiftingRequest.cy.ts:
```diff
@@ -0,0 +1,80 @@
+    shiftUpdate.typeAmbulanceNumber("2")
+    shiftUpdate.typeComment("New comment")
+    shiftUpdate.submitShiftForm();
+
```
remove unwanted spaces

nihal467 commented on cypress/pageobject/Shift/ShiftCreation.ts:
```diff
@@ -11,6 +11,22 @@ export class ShiftCreation {
+  }
+  
+  typeComment(comment:string) {
+    cy.get('#comments').click().type(comment)
```
```suggestion
    cy.get("#comments").click().type(comment)
```

nihal467 commented on cypress/pageobject/Shift/ShiftCreation.ts:
```diff
@@ -11,6 +11,22 @@ export class ShiftCreation {
+  }
+
   submitShiftForm() {
     cy.get("#submit").contains("Submit").click();
```
```suggestion
    cy.submitButton("Submit");
```

nihal467 commented on cypress/pageobject/Shift/ShiftUpdate.ts:
```diff
@@ -0,0 +1,27 @@
+    cy.get("#ambulance_driver_name").click().clear().type(name);
+  }
+  
+  typeAmbulancePhone(number: string) {
```
- you have created duplicate variable in two different files

Ayush1404 replied:
But the shiftUpdate class will require the function to be slightly different than the shiftCreation one , as they will clear the previous value that was already present in  the input box . That is why i created the fuctions seperately that might seem to be duplicate . If you want i can get rid of the functions but the clearing the original values wil have to  be done in original functions , is that okay ?

nihal467 replied:
don't create duplicate functions, check out the cy.typeIntoField() implementation in the code base, where we can use same variable for inputting and clearing by passing a boolean 

Make the usage and change across your test file

nihal467 commented on cypress/pageobject/Shift/ShiftingBoard.ts:
```diff
@@ -0,0 +1,8 @@
+        cy.get('#patient_name').type(patientName);
+        cy.intercept("GET", "**/api/v1/shift/**").as("getShiftingRequest");
+        cy.wait("@getShiftingRequest").its("response.statusCode").should("eq", 200);
+        cy.contains("All Details").click();
```
```suggestion
        cy.contains("All Details").click();
```

- give proper id for this button

Ayush1404 committed:
added requested changes

nihal467 commented:
@Ayush1404 what is the update ?

Ayush1404 commented:
@nihal467 I have made all the changes that you have suggested and pushed them , plz let me know if there are any other changes that are required

Ayush1404 commented:
Hey @nihal467 , are there any changes expected after i have made the last changes you suggested ?

nihal467 commented:
@coderabbitai full review

nihal467 commented:
@Ayush1404 fix the merge conflict and look into the code rabbit suggestion,

Ayush1404 commented:
@nihal467 I have removed the merge conflicts also fixed the change that code rabbit pointed out

nihal467 commented:
@Ayush1404 what is the status on this PR

Ayush1404 commented:
@nihal467 I rebased and removed merge conflict , but it is giving some importing errors on the build in PatientHone.tsx . I am not able to understand what happened .

Jacobjeevan commented:
> @nihal467 I rebased and removed merge conflict , but it is giving some importing errors on the build in PatientHone.tsx . I am not able to understand what happened .

PatientHome just changed, so probably a result of that. If there's import errors, probably an issue in merging 🤔 

Can you share screenshot of the errors? @Ayush1404

Ayush1404 commented:
> > @nihal467 I rebased and removed merge conflict , but it is giving some importing errors on the build in PatientHone.tsx . I am not able to understand what happened .
> 
> PatientHome just had a layout change, so probably a result of that. If there's import errors, probably an issue in merging 🤔
> 
> Can you share screenshot of the errors? @Ayush1404

Yes i will attach some screenshots , it is probably because PatientHome.tsx was completely changed . Should i just make another PR and close this one with the changes , cause as it was completely changed , the button i was adding id to doesn't even exist on that page , i think that will be better . 
<img width="1512" alt="Screenshot 2024-12-03 at 6 47 25 PM" src="https://github.com/user-attachments/assets/86aa74aa-e219-4685-80e7-5063ebbbb657">
<img width="1512" alt="Screenshot 2024-12-03 at 6 47 34 PM" src="https://github.com/user-attachments/assets/133831d7-e7c4-4b6e-a71a-8f3fb8872925">

Jacobjeevan commented:
> > > @nihal467 I rebased and removed merge conflict , but it is giving some importing errors on the build in PatientHone.tsx . I am not able to understand what happened .
> > 
> > 
> > PatientHome just had a layout change, so probably a result of that. If there's import errors, probably an issue in merging 🤔
> > Can you share screenshot of the errors? @Ayush1404
> 
> Yes i will attach some screenshots , it is probably because PatientHome.tsx was completely changed . Should i just make another PR and close this one with the changes , cause as it was completely changed , the button i was adding id to doesn't even exist on that page , i think that will be better . <img alt="Screenshot 2024-12-03 at 6 47 25 PM" width="1512" src="https://private-user-images.githubusercontent.com/59080746/391980645-86aa74aa-e219-4685-80e7-5063ebbbb657.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyMzQwMDEsIm5iZiI6MTczMzIzMzcwMSwicGF0aCI6Ii81OTA4MDc0Ni8zOTE5ODA2NDUtODZhYTc0YWEtZTIxOS00Njg1LTgwZTctNTA2M2ViYmJiNjU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjAzVDEzNDgyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBhN2I4NWU0MzJlZjE4N2U1Y2QwMGRkMDE2MWMxNzg0OTk0ODY3NDk2Y2JhZjA5ZWJhODM4YzljYzZmMGYwYmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.hJNVN09W0k6C25M3AsD_aWIgVK5uUkX97LaKLQ5xFEw"> <img alt="Screenshot 2024-12-03 at 6 47 34 PM" width="1512" src="https://private-user-images.githubusercontent.com/59080746/391980693-133831d7-e7c4-4b6e-a71a-8f3fb8872925.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyMzQwMDEsIm5iZiI6MTczMzIzMzcwMSwicGF0aCI6Ii81OTA4MDc0Ni8zOTE5ODA2OTMtMTMzODMxZDctZTdjNC00YjZlLWE3MWEtOGYzZmI4ODcyOTI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjAzVDEzNDgyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFjMGRmM2FmNjk3ZDY5YjhmNDk3YTk2MGRjZDk2YWM4ZDU5NzcwN2FlNzNkYTlmNzQ3MjVlZjlhYTU3MWM2YzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Xp-Xl446UwIeBMyr_QgUO16F-b_MOf9EIWWoGjXwYqg">

No, use this PR; use git to fetch/update the relevant file and adjust the tests accordingly.

nihal467 commented:
@Ayush1404 can you please share the current staus of the PR, otherwise in the next 48 hours, the PR will be closed and unassign you the issue.

Ayush1404 commented:
@nihal467  Hey , I have made the merge conflicts go away in the commit

nihal467 committed:
Merge branch 'develop' into issues/8914/shifting_request_test

nihal467 commented on cypress/e2e/shifting_spec/ShiftingRequest.cy.ts:
```diff
@@ -0,0 +1,94 @@
+  const currentFacilityPhone = "9465666768";
+  const shiftReasonInitial = "emmergency";
+  const ambulanceDriverNameInitial = "Rahul";
+  const ambulancePhoneInitial = "9865666768";
```
```suggestion
  const ambulancePhoneInitial = "9865666768";
```

use the generatePhoneNumber() utils function , don't use static phone number

nihal467 replied:
change it for all phone numbers

nihal467 commented on cypress/e2e/shifting_spec/ShiftingRequest.cy.ts:
```diff
@@ -0,0 +1,94 @@
+  const commentUpdated = "New comment";
+
+  before(() => { 
+    loginPage.loginAsDisctrictAdmin(); 
```
```suggestion
   loginPage.loginByRole("districtAdmin");
```

nihal467 commented on cypress/e2e/shifting_spec/ShiftingRequest.cy.ts:
```diff
@@ -0,0 +1,94 @@
+    cy.visit('/patients');
+    patientPage.visitPatient(patientName);
+    patientConsultationPage.clickShiftPatientButton();
+    
```
```suggestion
    remove unwanted space and add proper comments
```

nihal467 commented on cypress/e2e/shifting_spec/ShiftingRequest.cy.ts:
```diff
@@ -0,0 +1,94 @@
+    cy.wait("@shiftRequest").its("response.statusCode").should("eq", 201);
+
+    cy.visit("/shifting/board");
+    cy.contains(patientName).should("exist");
```
```suggestion
    cy.contains(patientName).should("exist");
```

- Create a id for the board card, then verify the content presence using cy.verifycontentpresence function
- Search for the patient name in the board first and using proper id, verify the patient card is visible in the board

nihal467 commented on cypress/e2e/shifting_spec/ShiftingRequest.cy.ts:
```diff
@@ -0,0 +1,94 @@
+      expect(responseData.ambulance_number).to.eq(ambulanceNumberUpdated);
+    });
+
+    cy.get('#shift_details_name').should('be.visible').and('have.text', patientName);
```
- use the command function for verifying content presence
- create a id for the card entirely where this data is present and verify all the variables together , that way you can remove line 83 to 88 and verify all content in a single line of code

nihal467 commented on cypress/pageobject/Patient/PatientConsultation.ts:
```diff
@@ -138,5 +138,14 @@ export class PatientConsultationPage {
+
+  clickShiftPatientButton() {
+    cy.get("#consultation-buttons").scrollIntoView();
+    cy.get("button").contains("Manage Patient").click();
```
```suggestion
    use clickManagePatientButton() from PatientConsultationPage
```

nihal467 replied:
reuse all existing function & always use proper element id's

nihal467 commented on cypress/pageobject/Patient/PatientConsultation.ts:
```diff
@@ -138,5 +138,14 @@ export class PatientConsultationPage {
+      "#consultation-buttons",
+      "Shift Patient",
+    );
+    cy.wait(3000);
```
```suggestion
    remove hardcorded wait and add proper cy.intercepts
```

nihal467 commented on cypress/pageobject/Shift/ShiftDetails.ts:
```diff
@@ -0,0 +1,5 @@
+export class ShiftDetails{
+    clickUpdateStatusButton(){
+        cy.contains("Update Status/Details").click()
```
```suggestion
        use submitButton or verifyandclick command function with proper id, dont try to interact with elements without proper id's
```

rithviknishad commented:
Closing due to inactivity
